refactor(cms): change pdf embed code

nickhsine opened the pull request:
Use mozilla/pdf.js lib to render pdf content.
Old way uses <iframe> to render pdf content, but it did not work well on mobile devices.
See [Asana Card](https://app.asana.com/0/1205517559975955/1208337869158726/f) for more info.

nickhsine committed:
refactor(cms): change pdf embed code
Use mozilla/pdf.js lib to render pdf content.
Old way uses <iframe> to render pdf content, but
it did not work well on mobile devices.
See [Asana
Card](https://app.asana.com/0/1205517559975955/1208337869158726/f) for more info.

duidae commented:
![Screen Shot 2024-10-01 at 4 13 30 PM](https://github.com/user-attachments/assets/529a3e3c-ba72-4d8a-8a57-3d3e18219788)

cant see pdf in safari ver 15.3

1. PDFjs could still have cross-browser issue according to stackoverflow
2. Instead of using embedded code, better to create a standalone component for pdf to avoid embedded code abuse.

nickhsine commented:
> 1. PDFjs could still have cross-browser issue according to stackoverflow

可以提供一下 stackoverflow 的連結嗎？另外，也想請你提供測試的頁面連結。

> 2. Instead of using embedded code, better to create a standalone component for pdf to avoid embedded code abuse.

這個可以討論看看之後要不要改用 draft button 來處理。
但因為時間關係，目前我會先用 embed code 的方式解決。

nickhsine commented:
你是用 https://dev-kids-cms.twreporter.org/preview-server/article/test-pdf 這篇文章在測 pdf 嗎？
如果是的話， embed code 裡面的檔案 url  的 hostname 要將 `dev-kids-cms.twreporter.org` 改成 `dev-kids-storage.twreporter.org`，`dev-kids-storage.twreporter.org` 的 response header 才有支援 CORS。

nickhsine commented:
想請你幫忙測試一下 https://dev-kids.twreporter.org/article/draft-test ，看看 pdf 是不是還是出不來？

duidae commented:
> 想請你幫忙測試一下 https://dev-kids.twreporter.org/article/draft-test ，看看 pdf 是不是還是出不來？
![Screen Shot 2024-10-04 at 1 32 54 PM](https://github.com/user-attachments/assets/b08bd611-0e4f-4f08-82b1-70b241fe984d)

[stackoverflow](https://stackoverflow.com/a/18704577)
https://kids-cms.twreporter.org/preview-server/article/test-pdf

nickhsine committed:
refactor(cms): downgrade pdfjs to 4.1.392 to support Safari 15.4+

Reference:
https://github.com/mozilla/pdf.js/blob/v4.1.392/gulpfile.mjs#L88

nickhsine committed:
refactor(cms): render pdf by <iframe> for older safari below 15.4

nickhsine committed:
fix(cms): fix pdfjs module path to support older Safari

nickhsine commented:
調整 pdfjs 的版本到 4.1.392，該版本支援 15.4+ 的 Safari 瀏覽器。

和偉婷討論的結果，因為 Safari 15 以下的使用者不多，所以用 iframe 當作 fallback。
其他瀏覽器和較高版本的 Safari 則用 pdfjs 呈現。

我這邊試了 15.4 以上的 Safari 瀏覽器，目前看起來 ok，再請 @duidae 幫忙看看 15.3 是否有畫面。

nickhsine committed:
fix(cms): remove `height: 100%` which restricts pdfjs to render whole pdf document

nickhsine committed:
fix(cms): correct pdf styles in iframe

duidae commented:
> 調整 pdfjs 的版本到 4.1.392，該版本支援 15.4+ 的 Safari 瀏覽器。
> 
> 和偉婷討論的結果，因為 Safari 15 以下的使用者不多，所以用 iframe 當作 fallback。 其他瀏覽器和較高版本的 Safari 則用 pdfjs 呈現。
> 
> 我這邊試了 15.4 以上的 Safari 瀏覽器，目前看起來 ok，再請 @duidae 幫忙看看 15.3 是否有畫面。

隨易測了三個都有問題: 無法顯示 or 只有第一頁

![Screen Shot 2024-10-08 at 4 59 17 PM](https://github.com/user-attachments/assets/b5215ef1-eea0-49ed-9db4-5663c111ef32)
![Screen Shot 2024-10-08 at 4 59 25 PM](https://github.com/user-attachments/assets/0895f3f0-731a-451e-811f-5fba58ad1e72)

![Screen Shot 2024-10-08 at 4 59 08 PM](https://github.com/user-attachments/assets/f46c7aef-e078-4d67-83ab-5899e77b61ec)

duidae reviewed (changes requested):
隨易測了三個都有問題: 無法顯示 or 只有第一頁

![Screen Shot 2024-10-08 at 4 59 17 PM](https://github.com/user-attachments/assets/b5215ef1-eea0-49ed-9db4-5663c111ef32)
![Screen Shot 2024-10-08 at 4 59 25 PM](https://github.com/user-attachments/assets/0895f3f0-731a-451e-811f-5fba58ad1e72)

![Screen Shot 2024-10-08 at 4 59 08 PM](https://github.com/user-attachments/assets/f46c7aef-e078-4d67-83ab-5899e77b61ec)

nickhsine commented:
想請問這個是用什麼模擬器測試的嗎？

我剛剛用了網路上免費的模擬器測了一下，看起來 iPhone 手機上的 Safari 16 版本以下的 pdfjs 還是會有問題，但 16 版本以上看起來 ok。
桌機（實體機）的部分則是 15.4 以上 ok。

我再問問看偉婷可不可接受 16 以下採用 fallback 的方式呈現 pdf。
或是我可能要再 downgrade pdfjs 的版本來試試看。
pdfjs 的版本只支援該版本近一年內的 Safari  版本和之後的版本。

nickhsine committed:
refactor(cms): update lists/pdf.ts
Rewrite CSS styles without nesting selector.

Nesting selector does not work on some older browsers.
For backwarads compatibility, rewrite them without nesting selector.

nickhsine committed:
refactor(cms): update lists/pdf.ts
Change the supported version of Safari from 15.4 to 16.

nickhsine commented:
和偉婷討論後，支援的 Safari 版本從 15.4 調整成 16 以上。

duidae commented:
> 和偉婷討論後，支援的 Safari 版本從 15.4 調整成 16 以上。

pdfjs在mobile上就是蠻高機率會遇到相容性問題, 靶歪了就調整靶, 好像有點把頭埋在沙子裡

nickhsine commented:
那我們再找看看其他的做法。

nickhsine commented:
我好像想到其他的做法了，我再試試看不要依據瀏覽器版本來判斷是否使用 pdfjs 還是 fallback 到 iframe 上。

nickhsine committed:
refactor(cms): render pdfjs if possible, otherwise fallback to iframe
Render iframe as default to support all kinds of browser. Only render pdfjs if browsers support it.

nickhsine committed:
refactor(cms): update lists/pdf.ts. Correct indent

nickhsine committed:
refactor(cms): update list/pdf.ts. Update comments

nickhsine committed:
refactor(cms): update lists/pdf.ts
`encodeURIComponent(pdfURL)`

nickhsine committed:
refactor(cms): update lists/pdf.ts. Remove Google Docs PDF viewer
use <iframe src="${pdfURL}"> instead.

In older browsers or devices, we are still encountering the issue where
the Google Docs PDF viewer shows a blank page in the iframe. The main
reason seems to be that Google Docs does not always return a 200
response, which would allow us to successfully view the PDF. Sometimes,
Google Docs returns a 204 response, leading to the iframe displaying a
blank page.

We currently suspect that the 204 response from Google Docs might be
related to the frequency of our requests.

nickhsine commented:
Commit [d5e36fc](https://github.com/kids-reporter/kids-reporter-monorepo/pull/671/commits/d5e36fc2f1721330fa00a337c0b7253cf8b7a133) 捨棄用 Safari 版本來決定要用什麼方式呈現 PDF 的邏輯，改成一律先用 iframe pdf URL 呈現，如果瀏覽器和裝置可以跑得動 pdfjs 的話，才改用 pdfjs 呈現。此 commit 可以解決白頁的問題。

註：
先前舊的瀏覽器或是裝置跑不動 pdfjs 的原因是 pdfjs 使用的 JavaScript 語法比較新，舊的無法支援。而 pdfjs 的團隊也有在文件說明他們的版本都只支援該版本前兩年的瀏覽器，而 Safari 更是只支援近一年的版本。
我目前是以 iOS 16（Safari 16）當做斷點，來選擇 pdfjs 的版本。

Commit [396e04d](https://github.com/kids-reporter/kids-reporter-monorepo/pull/671/commits/396e04d4b48cde2a24765273e7e99a40ae3f4256) 將 iframe pdf URL 改成用 Google Docs 來呈現 pdf。但透過第三方服務 [lambdatest](https://www.lambdatest.com/test-on-safari-browsers) 實測舊的裝置和瀏覽器，發現使用 Google Docs 來幫忙呈現 pdf 也是會遇到白頁的問題。經檢查，發現透過 iframe Google Docs 時，Google Docs 有時候會回傳 204 response（No content）而非 200 response（網頁內容），目前猜測可能是 Google Docs 有 rate limit 的機制，當 requests 量比較多時，Google Docs 會回傳 204 response。

因爲使用 Google Docs 也是會遇到白頁問題，所以和偉婷討論後，仍然改回 iframe pdf URL 的方式。
目前用 lambdatest 測試多組實機和瀏覽器，pdfjs 運作順利，舊的機款也都有使用 iframe pdf URL 的方式呈現。

所以我先 merge 此 PR 到 staging 上給偉婷近一步測試。

如果 PR 有任何建議，歡迎繼續留言在這個 thread。
